### PR TITLE
feat(fulltext_index): integrate full-text indexer with sst writer

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
#4246 

## What's changed and what's your intention?

* Add config for full-text index
* Integrate full-text index with sst writer
* TODO: tests til search available

## Checklist

- [x] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced new configuration options for full-text indexing in Mito engine, including settings for index creation, query application, and memory thresholds.

- **Bug Fixes**
  - Enhanced error handling capabilities for full-text index operations.

- **Documentation**
  - Updated configuration files with new full-text indexing options.

- **Tests**
  - Added new test cases to validate full-text indexing configurations and operations.

- **Chores**
  - Refined internal data structures and methods to support full-text indexing functionality.

### CODE REVIEW

**Comment by zhongzc** [comment]
@coderabbitai pause

**Review by fengjiachun — APPROVED** [review]
LGTM

File: config/datanode.example.toml [hunk 1]
```diff
@@ -434,6 +434,27 @@ mem_threshold_on_create = "64M"
 ## Deprecated, use `region_engine.mito.index.aux_path` instead.
 intermediate_path = ""
 
+## The options for full-text index in Mito engine.
+[region_engine.mito.fulltext_index]
+
+## Whether to create the index on flush.
```
Comment (killme2008) [inline config/datanode.example.toml:440]: Please add the default value of these options to the doc.

File: config/standalone.example.toml [hunk 2]
```diff
@@ -457,6 +457,27 @@ mem_threshold_on_create = "64M"
...
+## - `auto`: automatically
+## - `disable`: never
+create_on_compaction = "auto"
+
+## Whether to apply the index on query
+## - `auto`: automatically
+## - `disable`: never
+apply_on_query = "auto"
+
+## Memory threshold for index creation.
```
Comment (killme2008) [inline config/standalone.example.toml:478]: Is it too small? Maybe we can set the default value according to the system memory, just like 

https://github.com/GreptimeTeam/greptimedb/blob/09fff24ac4149a22f530a0e25df60c1aab0f5f73/src/mito2/src/config.rs#L145

Then the default value can be `auto`.

Comment (killme2008) [reply config/standalone.example.toml:478]: Other index memory or parallelism configurations can be tweaked this way too.

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        for column in &metadata.column_metadatas {
+            let options =
+                column
+                    .column_schema
+                    .fulltext_options()
+                    .context(FulltextOptionsSnafu {
+                        column_name: &column.column_schema.name,
+                    })?;
+
+            // Relax the type constraint here as many types can be casted to string.
```
Comment (killme2008) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:71]: I think we can constrain the type to be string, other types such as DateTime or integers, even though they can be cast into string, but the full text on for them is useless.

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        Ok(Self {
+            creators,
+            aborted: false,
+            stats: Statistics::default(),
+        })
+    }
+
+    /// Updates the index with the given batch.
+    pub async fn update(&mut self, batch: &Batch) -> Result<()> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
```
Comment (killme2008) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:113]: What about adding some metrics? About the timing and data size etc.

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        Ok(())
+    }
+
+    async fn do_finish(&mut self, puffin_writer: &mut SstPuffinWriter) -> Result<()> {
+        let mut guard = self.stats.record_finish();
+
+        let mut written_bytes = 0;
+        for creator in self.creators.values_mut() {
+            written_bytes += creator.finish(puffin_writer).await?;
+        }
```
Comment (killme2008) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:193]: Prefer the iterator with reduce and sum, but it's ok.

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+        guard.inc_byte_count(written_bytes);
+        Ok(())
+    }
+
+    async fn do_abort(&mut self) -> Result<()> {
+        let _guard = self.stats.record_cleanup();
+
```
Comment (killme2008) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:201]: Do we need to add a check?

```rust
if self.aborted {
     return Ok(());
}
```

File: src/mito2/src/sst/index/fulltext_index/creator.rs [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+                    self.inner
+                        .push_text(text)
+                        .await
+                        .context(FulltextPushTextSnafu)?;
+                }
+            }
+            _ => {
+                for _ in 0..batch.num_rows() {
+                    self.inner
+                        .push_text("")
```
Comment (killme2008) [inline src/mito2/src/sst/index/fulltext_index/creator.rs:255]: Not quite sure, in such a case,do we still need to index the empty string?

File: src/mito2/src/sst/index/indexer/finish.rs [hunk 8]
```diff
@@ -36,10 +38,43 @@ impl Indexer {
...
+        }
+
         output.file_size = self.do_finish_puffin_writer(writer).await;
         output
     }
 
+    async fn build_puffin_writer(&mut self) -> Option<SstPuffinWriter> {
+        let Some(puffin_manager) = self.puffin_manager.take() else {
+            return None;
+        };
```
Comment (killme2008) [inline src/mito2/src/sst/index/indexer/finish.rs:56]:
```suggestion
        let puffin_manager) = self.puffin_manager.take();
```

File: src/mito2/src/sst/index.rs [hunk 9]
```diff
@@ -117,31 +143,28 @@ pub(crate) struct IndexerBuilder<'a> {
...
-        if indexer.inverted_indexer.is_none() {
-            indexer.abort().await;
-            return Indexer::default();
-        }
-
-        indexer.puffin_writer = self.build_puffin_writer().await;
-        if indexer.puffin_writer.is_none() {
-            indexer.abort().await;
+        indexer.fulltext_indexer = self.build_fulltext_indexer().await;
+        if indexer.inverted_indexer.is_none() && indexer.fulltext_indexer.is_none() {
```
Comment (killme2008) [inline src/mito2/src/sst/index.rs:163]: Does the `indexer` need to abort before returning just like before?

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 5 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:193]: Clause with async and &mut ref is annoying. The following code is closed but still doesn't work due to ref of `puffin_writer`.

```rust
let written_bytes = stream::iter(self.creators.values_mut())
    .then(|creator| async move { creator.finish(puffin_writer).await })
    .try_fold(0, |acc, x| async move { Ok(acc + x) })
    .await?;
```

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 4 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:113]: In do_update, `stats` will record the metrics and flush when dropped. However, you reminded me that it is possible to distinguish between the metrics of the inverted index and the full-text index.

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 3 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:71]: My idea is that specific constraints can be implemented by the frontend, and by relaxing constraints in the mito, many things can function.

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 6 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:201]: Agreed, although this `abort` doesn't do much, it reduces the mental burden for the code reader, so I will add it.

File: src/mito2/src/sst/index.rs (hunk 9 above)
Comment (zhongzc) [reply src/mito2/src/sst/index.rs:163]: Agreed, although this `abort` doesn't do much, it reduces the mental burden for the code reader, so I will add it.

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 7 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:255]: If the `text` column in a certain batch is all `None`, we still need to `push("")` to ensure the row numbers in the index match.

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 6 above)
Comment (zhongzc) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:201]: https://github.com/GreptimeTeam/greptimedb/blob/21b38311d777cf931e5ffe55653577f5448f30ec/src/mito2/src/sst/index/fulltext_index/creator.rs#L153-L157

`abort` has checked it.

File: config/standalone.example.toml (hunk 2 above)
Comment (zhongzc) [reply config/standalone.example.toml:478]: good idea!

Comment (zhongzc) [reply config/standalone.example.toml:480]: updated

File: src/mito2/src/sst/index/fulltext_index/creator.rs (hunk 7 above)
Comment (killme2008) [reply src/mito2/src/sst/index/fulltext_index/creator.rs:255]: @zhongzc Got it, could you add a comment for it? Other people may have the same question.

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**